{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "        <script type=\"text/javascript\">\n",
       "        window.PlotlyConfig = {MathJaxConfig: 'local'};\n",
       "        if (window.MathJax) {MathJax.Hub.Config({SVG: {font: \"STIX-Web\"}});}\n",
       "        if (typeof require !== 'undefined') {\n",
       "        require.undef(\"plotly\");\n",
       "        requirejs.config({\n",
       "            paths: {\n",
       "                'plotly': ['https://cdn.plot.ly/plotly-latest.min']\n",
       "            }\n",
       "        });\n",
       "        require(['plotly'], function(Plotly) {\n",
       "            window._Plotly = Plotly;\n",
       "        });\n",
       "        }\n",
       "        </script>\n",
       "        "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import sys\n",
    "import argparse\n",
    "import numpy as np\n",
    "from scipy.stats import chi2_contingency\n",
    "from itertools import product\n",
    "import itertools as it\n",
    "\n",
    "from _plotly_future_ import v4_subplots\n",
    "import plotly.graph_objs as go\n",
    "from plotly.offline import download_plotlyjs, init_notebook_mode, plot, iplot\n",
    "from plotly.graph_objs import *\n",
    "\n",
    "from ipywidgets import interact, interactive, fixed, interact_manual\n",
    "import ipywidgets as widgets\n",
    "init_notebook_mode(connected=True)\n",
    "\n",
    "from datetime import datetime\n",
    "import tempfile\n",
    "import os\n",
    "import gzip \n",
    "import subprocess\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Sim pipeline\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "from tools.SLiM_pipe_tools import (\n",
    "    read_chrom_sizes, region_sample, region_samplev1,\n",
    "    fasta_Rextract, write_fastaEx, process_recipe,\n",
    "    SLiM_dispenser, \n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/mnt/d/GitHub/fine-scale-mutation-spectrum-master/slim_pipe\n"
     ]
    }
   ],
   "source": [
    "print(os.getcwd())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "## directories\n",
    "main_dir= '/mnt/d/GitHub/fine-scale-mutation-spectrum-master/slim_pipe/'\n",
    "slim_dir= '/mnt/d/GitHub/fine-scale-mutation-spectrum-master/slim_pipe/'\n",
    "fastas_dir= '/mnt/d/GitHub/fine-scale-mutation-spectrum-master/slim_pipe/Fastas/'\n",
    "##\n",
    "\n",
    "## sub-directories.\n",
    "sim_dir= main_dir + 'Recipes/Human_sims/'\n",
    "sim_recipe= 'Gravel_2011_kmmVCF_fasta_pipe.slim'\n",
    "\n",
    "dir_data= main_dir + 'mutation_counter/data/'\n",
    "dir_vcf= dir_data + 'vcf_data/'\n",
    "count_dir= main_dir + 'mutation_counter/count/'\n",
    "dir_launch= main_dir + 'mutation_counter'\n",
    "\n",
    "summary_file= 'sims.log'\n",
    "\n",
    "#\n",
    "##\n",
    "##\n",
    "slim_soft= slim_dir + 'sim*'\n",
    "sim_recipe= sim_dir + sim_recipe\n",
    "##\n",
    "##\n",
    "#"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Simulation tag names, assembly to select from.\n",
    "batch_name= 'Gravel'\n",
    "assembly= 'hg38'\n",
    "\n",
    "## files & variables\n",
    "## fasta segment lengths; number of segments / sims.\n",
    "L= int(1e4)\n",
    "N= 2\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "def region_samplev2(L, chrom_sizes, N, fasta_file= ''):\n",
    "    '''\n",
    "    prepare sequence dictionary: {chrom:list(start pos)}\n",
    "    provide to fasta_Rextract.\n",
    "    \n",
    "    ##\n",
    "    '''\n",
    "    \n",
    "    chroms= list(chrom_sizes.keys())\n",
    "    sizes= [int(chrom_sizes[x]) for x in chroms]\n",
    "    sizes= np.array(sizes) / sum(sizes)\n",
    "    \n",
    "    choices= np.random.choice(chroms,size= N,p= sizes,replace= True)\n",
    "    print(choices)\n",
    "    \n",
    "    seqs_store= {\n",
    "        z: len([x for x in range(N) if choices[x]==z]) for z in list(set(choices))\n",
    "    }\n",
    "    \n",
    "    seqs= fasta_RextractUnif(fasta_file, seqs_store, L= L)\n",
    "    \n",
    "    return seqs\n",
    "\n",
    "\n",
    "\n",
    "def fasta_RextractUnif(fasta,seq_store,L= 10000):\n",
    "    ''' \n",
    "    Extract regions from fasta file.\n",
    "    Takes dictionary {chrom: list(start points)};\n",
    "    extracts sequences after reading each chromosome.\n",
    "    '''\n",
    "    refseqs= {x:{} for x in seq_store.keys()}\n",
    "    \n",
    "    Outfiles= {\n",
    "        x: tempfile.TemporaryFile() for x in seq_store.keys()\n",
    "    }\n",
    "    \n",
    "    d= 0\n",
    "     \n",
    "    with gzip.open(fasta,'rb') as fp:\n",
    "        for line in fp:\n",
    "            #line= line.decode()\n",
    "            if line[0:1] == b'>':\n",
    "                head= line.decode()\n",
    "                head= head.strip()\n",
    "                if d != 0:\n",
    "                    d=0\n",
    "\n",
    "                head= head[1:].split('\\t')\n",
    "                head= head[0].strip('chr')\n",
    "\n",
    "                if head in seq_store.keys():\n",
    "                    print('opening fasta chr: {}'.format(head))\n",
    "                    d= head\n",
    "                    outfile= Outfiles[d]\n",
    "                    continue\n",
    "            \n",
    "            if d != 0:\n",
    "                processed_line= line.upper().strip(b'\\n')\n",
    "                outfile.write(processed_line)\n",
    "    \n",
    "    for chrom in Outfiles.keys():\n",
    "        f= Outfiles[chrom]\n",
    "        f.seek(os.SEEK_SET)\n",
    "        result= f.read().decode()\n",
    "        f.close()\n",
    "        \n",
    "        chrom_seqs= return_seqs(result,size= seq_store[chrom],L= L)\n",
    "        \n",
    "        refseqs[chrom]= chrom_seqs\n",
    "    \n",
    "    return refseqs\n",
    "\n",
    "\n",
    "\n",
    "def return_seqs(seq,size= 10,L= 1000,keep= ['A','T','G','C']):\n",
    "    \n",
    "    d= 0\n",
    "    \n",
    "    seqL= len(seq)\n",
    "    seq_dict= {}\n",
    "    \n",
    "    while d <= size:\n",
    "        pos= np.random.randint(low= 0, high= seqL - L,size= 1)[0]\n",
    "        given= seq[pos:(pos + L)]\n",
    "        \n",
    "        scrag= [x for x in given if x not in keep]\n",
    "        \n",
    "        if not scrag:\n",
    "            seq_dict[pos]= given\n",
    "            \n",
    "            d += 1\n",
    "    \n",
    "    return seq_dict\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['7' '22']\n",
      "opening fasta chr: 22\n",
      "opening fasta chr: 7\n",
      "AACCATTATA\n",
      "AAAAAATTAG\n",
      "AAATGTCAAC\n",
      "ATTTAATACA\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "\n",
    "## Read chrom_sizes file to decide where to sample files from. \n",
    "chrom_sizes= read_chrom_sizes(assembly)\n",
    "\n",
    "## Sample fasta.\n",
    "##\n",
    "fasta= fastas_dir + assembly + '.fa.gz'\n",
    "rseqs= region_samplev2(L, chrom_sizes,N, fasta)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def cook_constants_v1(fasta_dict, dir_data= \"./data/\", \n",
    "            dir_vcf= \"vcf_data/\", slim_dir= './', batch_name= ''):\n",
    "    '''\n",
    "    set up conditions.\n",
    "    constants:\n",
    "        - vcf_file;\n",
    "        - fasta_file - writes fasta; mkdir fasta_dir\n",
    "    '''\n",
    "    sim_store= {}\n",
    "    \n",
    "    for chrom in fasta_dict.keys():\n",
    "        for start in fasta_dict[chrom].keys():\n",
    "            fasta= fasta_dict[chrom][start]\n",
    "\n",
    "            ### set up names and directories.\n",
    "            SIMname= batch_name + 'C{}.{}'.format(chrom,str(start))\n",
    "            vcf_file= dir_vcf + SIMname + \"_chr{}.vcf\".format(chrom)\n",
    "            ref_dir= dir_data + SIMname + '_reference'\n",
    "            os.makedirs(ref_dir, exist_ok=True)\n",
    "            \n",
    "            ### write fasta file for SLiM.\n",
    "            fasta_file= write_fastaEx(fasta,chrom=chrom,start= start,\n",
    "                          ID= SIMname,fasta_dir= ref_dir)\n",
    "            \n",
    "            sim_store[SIMname]= {\n",
    "                \"vcf_file\": vcf_file,\n",
    "                \"fasta_file\": fasta_file\n",
    "            }\n",
    "    \n",
    "    return sim_store\n",
    "\n",
    "\n",
    "def cook_constants_v2(fasta_dict, dir_data= \"./data/\", \n",
    "            dir_vcf= \"vcf_data/\", slim_dir= './', batch_name= ''):\n",
    "    '''\n",
    "    set up conditions.\n",
    "    constants:\n",
    "        - vcf_file;\n",
    "        - fasta_file - writes fasta; mkdir fasta_dir\n",
    "    '''\n",
    "    sim_store= {}\n",
    "    \n",
    "    for chrom in fasta_dict.keys():\n",
    "        for start in fasta_dict[chrom].keys():\n",
    "            fasta= fasta_dict[chrom][start]\n",
    "\n",
    "            ### set up names and directories.\n",
    "            SIMname= batch_name + 'C{}.{}'.format(chrom,str(start))\n",
    "            vcf_file= dir_vcf + SIMname + \"_chr{}.vcf\".format(chrom)\n",
    "            ref_dir= dir_data + SIMname + '_reference'\n",
    "            os.makedirs(ref_dir, exist_ok=True)\n",
    "            \n",
    "            ### write fasta file for SLiM.\n",
    "            fasta_file= write_fastaEx(fasta,chrom=chrom,start= start,\n",
    "                          ID= SIMname,fasta_dir= ref_dir)\n",
    "            \n",
    "            sim_store[SIMname]= {\n",
    "                \"vcf_file\": vcf_file,\n",
    "                \"fasta_file\": fasta_file\n",
    "            }\n",
    "    \n",
    "    return sim_store\n",
    "\n",
    "\n",
    "def SLiM_dispenserv1(sim_store, slim_dir= './', batch_name= '',\n",
    "                    logSims= 'sims.log', mutlog= 'toMut.log'):\n",
    "    ''' execute SLiM program\n",
    "    - simulation specific recipe:\n",
    "    - recipe template is re-written to direct to new fasta.\n",
    "    '''\n",
    "    nf= len(sim_store)\n",
    "    for SIMname in sim_store.keys():\n",
    "        \n",
    "        command_line_constants= sim_store[SIMname]\n",
    "\n",
    "        ### generate modified slim recipe\n",
    "        new_recipe= process_recipe(sim_recipe,command_line_constants, SIMname)\n",
    "\n",
    "        seed= np.random.randint(0,high= nf,size= 1)[0]\n",
    "        ### Launch SLiM through shell.\n",
    "        slim_soft= slim_dir + 'slim*' \n",
    "\n",
    "        command_units= [slim_soft, '-m', '-s', str(seed), new_recipe]\n",
    "\n",
    "        os.system(' '.join(command_units))\n",
    "\n",
    "        os.system('gzip {}'.format(sim_store[SIMname][\"vcf_file\"]))\n",
    "        os.system('gzip {}'.format(sim_store[SIMname][\"fasta_file\"]))\n",
    "        os.system('rm {}'.format(new_recipe))\n",
    "\n",
    "        now = datetime.now()\n",
    "        tnow= now.strftime(\"%d/%m/%Y %H:%M:%S\")\n",
    "\n",
    "        with open(logSims,'a') as fp:\n",
    "            fp.write(SIMname + '\\t' + tnow + '\\n')\n",
    "        with open(mutlog,'a') as fp:\n",
    "            fp.write(SIMname + '\\n')\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [],
   "source": [
    "mutlog= 'toMut.log'\n",
    "## Perform Simulations\n",
    "print('launch SLiM jobs.')\n",
    "\n",
    "sim_store= cook_constants_v1(rseqs,dir_data= dir_data, dir_vcf= dir_vcf, \n",
    "               slim_dir= slim_dir, batch_name= batch_name)\n",
    "\n",
    "SLiM_dispenserv1(sim_store, slim_dir= slim_dir, batch_name= batch_name,\n",
    "                    logSims= summary_file, mutlog= mutlog)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "SLiM_dispenser(rseqs,sim_recipe,dir_data= dir_data, dir_vcf= dir_vcf, \n",
    "               slim_dir= slim_dir, logSims= summary_file,\n",
    "               mutlog= mutlog,batch_name= batch_name)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "launch mutation counter.\n",
      "['GravelC7.145319599\\n', 'GravelC7.10890494\\n', 'GravelC22.40861331\\n', 'GravelC22.46101549\\n']\n"
     ]
    }
   ],
   "source": [
    "from tools.SLiM_pipe_tools import mutation_counter_launch\n",
    "\n",
    "mutlog= 'toMut.log'\n",
    "\n",
    "print('launch mutation counter.')\n",
    "mutation_counter_launch(mutlog,count_dir= count_dir, \n",
    "                        dir_launch= dir_launch,main_dir= main_dir)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
